experimental/yarpl/src/yarpl/flowable/sources/Flowable_RangeSubscription.cpp

```diff
@@ -0,0 +1,102 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
+
+#include "Flowable_RangeSubscription.h"
+#include <iostream>
+#include "yarpl/flowable/utils/SubscriptionHelper.h"
+
+namespace yarpl {
+namespace flowable {
+namespace sources {
+
+using namespace yarpl::flowable::internal;
+
+RangeSubscription::RangeSubscription(
+    long start,
+    long count,
+    std::unique_ptr<Subscriber> subscriber)
+    : current_(start),
+      max_(start + count - 1),
+      subscriber_(std::move(subscriber)){};
+
+void RangeSubscription::start() {
+  subscriber_->onSubscribe(this);
+}
+
+void RangeSubscription::request(int64_t n) {
+  if (n <= 0) {
+    return;
+  }
+  int64_t r = SubscriptionHelper::addCredits(&requested_, n);
+  if (r <= 0) {
+    return;
+  }
+  auto consumed = tryEmit();
+  if (consumed > 0) {
+    SubscriptionHelper::consumeCredits(&requested_, consumed);
+
+    // if we've hit the end, or been cancelled, delete ourselves
+    // it will be >max_ since we suffix increment i.e. current_++
+    if (current_ > max_) {
+      subscriber_->onComplete();
+      tryDelete();
+      return;
+    } else if (SubscriptionHelper::isCancelled(&requested_)) {
+      tryDelete();
+      return;
+    }
+  }
+}
+
+/** do the actual work of emission */
+int64_t RangeSubscription::tryEmit() {
+  int64_t consumed{0};
+  int64_t r{0};
```

Comment:
nit: you can declare r in the for( auto r = reuqested_.load() ; ... ) init section